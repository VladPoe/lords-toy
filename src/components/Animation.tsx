import React, { useEffect, useState } from "react";
import styled, {
  FlattenSimpleInterpolation,
  keyframes,
  Keyframes,
  css,
} from "styled-components";
import { KeysFromConst } from "../utils/types";

export const ANIMATION_TYPES = {
  FADE_IN: "fadeId",
  TREMBLE: "tremble",
  FROM_LEFT: "fromLeft",
} as const;

export type TAnimation = KeysFromConst<typeof ANIMATION_TYPES>;

const KEYFRAMES: Record<TAnimation, Keyframes> = {
  [ANIMATION_TYPES.FADE_IN]: keyframes`
    0% {
      opacity: 0;
      transform: scale(0);
    }
    60% {
      opacity: 0.3;
      transform: scale(0.6);
    }
    80% {
      opacity: 0.5;
      transform: scale(0.8);
    }
    100% {
      opacity: 1;
      transform: scale(1);
    }
  `,
  [ANIMATION_TYPES.TREMBLE]: keyframes`
    0% {
      transform: translate(0, 0);
    }
    10% {
      transform: translate(2px, -2px);
    }
    20% {
      transform: translate(0, 0);
    }
    30% {
      transform: translate(-2px, -2px);
    }
    40% {
      transform: translate(1px, -1px);
    }
    50% {
      transform: translate(0, 0);
    }
    60% {
      transform: translate(-3px, -3px);
    }
    70% {
      transform: translate(0, 0);
    }
    80% {
      transform: translate(2px, -2px);
    }
    90% {
      transform: translate(-1px, -1px);
    }
    100% {
      transform: translate(0, 0);
    }
  `,
  [ANIMATION_TYPES.FROM_LEFT]: keyframes`
    0% {
      transform: translate(-100%, 0);
      opacity: 0;
    }
    50% {
      transform: translate(4px, 0);
      opacity: 1
    }
    100% {
      transform: translate(0, 0);
    }
  `,
};

const ANIMATION: Record<TAnimation, FlattenSimpleInterpolation> = {
  [ANIMATION_TYPES.FADE_IN]: css`
    animation: 400ms ease-in ${KEYFRAMES[ANIMATION_TYPES.FADE_IN]};
  `,
  [ANIMATION_TYPES.TREMBLE]: css`
    animation: 500ms linear ${KEYFRAMES[ANIMATION_TYPES.TREMBLE]};
  `,
  [ANIMATION_TYPES.FROM_LEFT]: css`
    animation: 400ms ease-in ${KEYFRAMES[ANIMATION_TYPES.FROM_LEFT]};
  `,
};

const AnimationWrap = styled.div<{ animated: boolean; type: TAnimation }>`
  ${({ type }) => ANIMATION[type]};
`;

export const Animation: React.FC<{ type: TAnimation }> = ({
  type,
  children,
}) => {
  const [animated, setAnimatedStats] = useState(false);
  useEffect(() => {
    requestAnimationFrame(() => {
      setAnimatedStats(true);
    });
  }, []);
  return (
    <AnimationWrap animated={animated} type={type}>
      {children}
    </AnimationWrap>
  );
};
