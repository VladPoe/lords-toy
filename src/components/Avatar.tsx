import React from "react";
import styled, { css, FlattenSimpleInterpolation } from "styled-components";
import { KeysFromConst } from "../utils/types";
import Skull from "../assets/images/skull.png";
import Chicken from "../assets/images/chicken.png";
import Flash from "../assets/images/flash.png";

export const AVATARS = {
  ALIVE_CELL: "aliveCell",
  DEAD_CELL: "deadCell",
  LIFE: "life",
} as const;

export type TAvatar = KeysFromConst<typeof AVATARS>;

export const GRADIENT_MAP: Record<TAvatar, string> = {
  [AVATARS.DEAD_CELL]: "linear-gradient(180deg,#0D658A 0%,#B0FFB4 100%)",
  [AVATARS.ALIVE_CELL]: "linear-gradient(180deg,#FFB800 0%,#FFF7B0 100%)",
  [AVATARS.LIFE]: "linear-gradient(180deg,#AD00FF 0%,#FFB0E9 100%)",
};

const IMAGES_MAP: Record<TAvatar, string> = {
  [AVATARS.DEAD_CELL]: Skull,
  [AVATARS.ALIVE_CELL]: Flash,
  [AVATARS.LIFE]: Chicken,
};

const ALT_TITLES_MAP: Record<TAvatar, string> = {
  [AVATARS.DEAD_CELL]: "skull",
  [AVATARS.ALIVE_CELL]: "flash",
  [AVATARS.LIFE]: "chicken",
};

const ALIGNING: Record<TAvatar, FlattenSimpleInterpolation> = {
  [AVATARS.DEAD_CELL]: css`
    transform: translate(-1px, 0px);
  `,
  [AVATARS.ALIVE_CELL]: css`
    transform: translate(-2px, -1px);
  `,
  [AVATARS.LIFE]: css`
    transform: translate(-1px, 0px);
  `,
};

const AvatarContainer = styled.div<{ type: TAvatar }>`
  width: 40px;
  height: 40px;
  border-radius: 50%;
  ${({ type }) => `background-image: ${GRADIENT_MAP[type]};`}
  display: inline-flex;
  align-items: center;
  justify-content: center;
`;

const Img = styled.img<{ type: TAvatar }>`
  width: 21px;
  height: 21px;
  ${({ type }) => ALIGNING[type]}
`;

export const Avatar: React.FC<{ type: TAvatar }> = ({ type }) => {
  return (
    <AvatarContainer type={type}>
      <Img
        type={type}
        alt={ALT_TITLES_MAP[type]}
        width="21px"
        height="auto"
        src={IMAGES_MAP[type]}
      />
    </AvatarContainer>
  );
};
