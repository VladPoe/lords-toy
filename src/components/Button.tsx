import React from "react";
import styled from "styled-components";
import { violet, violetBlack, yellow } from "../tokens/colors";
import { adaptive } from "../tokens/screen";

const ButtonStyled = styled.button`
  cursor: pointer;
  color: white;
  background-color: ${violet};
  padding-top: 10px;
  padding-bottom: 11px;
  text-transform: uppercase;
  text-align: center;
  letter-spacing: 0.0125em;
  width: 100%;
  border: none;
  border-radius: 4px;
  font-size: 14px;
  font-weight: 500;
  transition-property: background-color, color;
  transition-duration: 200ms;

  &:hover {
    background-color: ${yellow};
    color: ${violetBlack};
  }

  ${adaptive.M} {
    padding-top: 18px;
    padding-bottom: 18px;
  }
`;

export type TButton = {
  text: string;
  htmlTitle?: string;
  type?: "button" | "submit" | "reset";
  onClick: React.MouseEventHandler<HTMLButtonElement>;
};

export const Button: React.FC<TButton> = ({
  text,
  htmlTitle,
  type = "button",
  onClick,
}) => {
  return (
    <ButtonStyled title={htmlTitle} type={type} onClick={onClick}>
      {text}
    </ButtonStyled>
  );
};
