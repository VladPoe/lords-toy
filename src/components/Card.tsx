import React from "react";
import styled from "styled-components";

export type TCard = {
  title: string;
  description: string;
  iconMark?: React.ReactElement;
};

const AvatarWrap = styled.div``;

const InfoWrap = styled.div``;

const Title = styled.h6`
  font-weight: 500;
  font-size: 20px;
  margin: 0;
  letter-spacing: 0.0015em;
`;

const Description = styled.span`
  letter-spacing: 0.0025em;
  display: block;
`;

const CardContainer = styled.div`
  padding: 16px;
  background-color: white;
  border-radius: 8px;
  display: flex;

  ${AvatarWrap} + ${InfoWrap} {
    margin-left: 16px;
  }

  ${Title} + ${Description} {
    margin-top: 3px;
  }
`;

export const Card: React.FC<TCard> = ({ title, description, iconMark }) => {
  return (
    <CardContainer>
      {iconMark && <AvatarWrap>{iconMark}</AvatarWrap>}
      <InfoWrap>
        <Title>{title}</Title>
        <Description>{description}</Description>
      </InfoWrap>
    </CardContainer>
  );
};
