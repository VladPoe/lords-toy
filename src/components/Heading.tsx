import React from "react";
import styled from "styled-components";

export const Heading = styled.h1`
  color: white;
  margin: 0;
  text-align: center;
  letter-spacing: 0.0015em;
  line-height: 28px;
  font-weight: 500;
  font-size: 20px;
`;
