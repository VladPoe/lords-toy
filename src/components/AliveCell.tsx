import React from "react";
import { Card } from "./Card";
import { AVATARS, Avatar } from "./Avatar";
import { Animation, ANIMATION_TYPES } from "./Animation";

const TITLE = "Живая";
const DESCRIPTION = "и шевелится!";

export const AliveCell: React.FC = () => {
  return (
    <Animation type={ANIMATION_TYPES.FROM_LEFT}>
      <Card
        title={TITLE}
        description={DESCRIPTION}
        iconMark={<Avatar type={AVATARS.ALIVE_CELL} />}
      />
    </Animation>
  );
};
