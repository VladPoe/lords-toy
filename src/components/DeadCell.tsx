import React from "react";
import { Card } from "./Card";
import { AVATARS, Avatar } from "./Avatar";
import { Animation, ANIMATION_TYPES } from "./Animation";

const TITLE = "Мёртвая";
const DESCRIPTION = "или прикидывается";

export const DeadCell: React.FC = () => {
  return (
    <Animation type={ANIMATION_TYPES.TREMBLE}>
      <Card
        title={TITLE}
        description={DESCRIPTION}
        iconMark={<Avatar type={AVATARS.DEAD_CELL} />}
      />
    </Animation>
  );
};
