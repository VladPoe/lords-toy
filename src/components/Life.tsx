import React from "react";
import { Card } from "./Card";
import { AVATARS, Avatar } from "./Avatar";
import { Animation, ANIMATION_TYPES } from "./Animation";

const TITLE = "Жизнь";
const DESCRIPTION = "Ку-ку!";

export const Life: React.FC = () => {
  return (
    <Animation type={ANIMATION_TYPES.FADE_IN}>
      <Card
        title={TITLE}
        description={DESCRIPTION}
        iconMark={<Avatar type={AVATARS.LIFE} />}
      />
    </Animation>
  );
};
