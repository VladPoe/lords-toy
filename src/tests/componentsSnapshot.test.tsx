import "jest-styled-components";
import React from "react";
import renderer from "react-test-renderer";

import { AliveCell } from "../components/AliveCell";
import { DeadCell } from "../components/DeadCell";
import { Life } from "../components/Life";
import { Button } from "../components/Button";

describe("Basic visual component should be rendered correctly", () => {
  it("AliveCell valid snapshot", () => {
    const aliveCell = renderer.create(<AliveCell />).toJSON();
    expect(aliveCell).toMatchSnapshot();
  });

  it("DeadCell valid snapshot", () => {
    const deadCell = renderer.create(<DeadCell />).toJSON();
    expect(deadCell).toMatchSnapshot();
  });

  it("Life component valid snapshot", () => {
    const life = renderer.create(<Life />).toJSON();
    expect(life).toMatchSnapshot();
  });

  it("Button component valid snapshot", () => {
    const button = renderer
      .create(<Button text="test button" type="submit" onClick={() => null} />)
      .toJSON();
    expect(button).toMatchSnapshot();
  });
});
