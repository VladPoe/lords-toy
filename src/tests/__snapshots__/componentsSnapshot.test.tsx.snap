// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic visual component should be rendered correctly AliveCell valid snapshot 1`] = `
.c5 {
  font-weight: 500;
  font-size: 20px;
  margin: 0;
  -webkit-letter-spacing: 0.0015em;
  -moz-letter-spacing: 0.0015em;
  -ms-letter-spacing: 0.0015em;
  letter-spacing: 0.0015em;
}

.c7 {
  -webkit-letter-spacing: 0.0025em;
  -moz-letter-spacing: 0.0025em;
  -ms-letter-spacing: 0.0025em;
  letter-spacing: 0.0025em;
  display: block;
}

.c1 {
  padding: 16px;
  background-color: white;
  border-radius: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 .+ .{
  margin-left: 16px;
}

.c1 .c4 + .c6 {
  margin-top: 3px;
}

.c2 {
  width: 40px;
  height: 40px;
  border-radius: 50%;
  background-image: linear-gradient(180deg,#FFB800 0%,#FFF7B0 100%);
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  width: 21px;
  height: 21px;
  -webkit-transform: translate(-2px,-1px);
  -ms-transform: translate(-2px,-1px);
  transform: translate(-2px,-1px);
}

.c0 {
  -webkit-animation: 400ms ease-in fLgObN;
  animation: 400ms ease-in fLgObN;
}

<div
  className="c0"
  type="fromLeft"
>
  <div
    className="c1"
  >
    <div
      className=""
    >
      <div
        className="c2"
        type="aliveCell"
      >
        <img
          alt="flash"
          className="c3"
          height="auto"
          src="flash.png"
          type="aliveCell"
          width="21px"
        />
      </div>
    </div>
    <div
      className=""
    >
      <h6
        className="c4 c5"
      >
        Живая
      </h6>
      <span
        className="c6 c7"
      >
        и шевелится!
      </span>
    </div>
  </div>
</div>
`;

exports[`Basic visual component should be rendered correctly Button component valid snapshot 1`] = `
.c0 {
  cursor: pointer;
  color: white;
  background-color: #5A3472;
  padding-top: 10px;
  padding-bottom: 11px;
  text-transform: uppercase;
  text-align: center;
  -webkit-letter-spacing: 0.0125em;
  -moz-letter-spacing: 0.0125em;
  -ms-letter-spacing: 0.0125em;
  letter-spacing: 0.0125em;
  width: 100%;
  border: none;
  border-radius: 4px;
  font-size: 14px;
  font-weight: 500;
  -webkit-transition-property: background-color,color;
  transition-property: background-color,color;
  -webkit-transition-duration: 200ms;
  transition-duration: 200ms;
}

.c0:hover {
  background-color: #ffc230;
  color: #09000e;
}

@media only screen and (min-width:540px) {
  .c0 {
    padding-top: 18px;
    padding-bottom: 18px;
  }
}

<button
  className="c0"
  onClick={[Function]}
  type="submit"
>
  test button
</button>
`;

exports[`Basic visual component should be rendered correctly DeadCell valid snapshot 1`] = `
.c5 {
  font-weight: 500;
  font-size: 20px;
  margin: 0;
  -webkit-letter-spacing: 0.0015em;
  -moz-letter-spacing: 0.0015em;
  -ms-letter-spacing: 0.0015em;
  letter-spacing: 0.0015em;
}

.c7 {
  -webkit-letter-spacing: 0.0025em;
  -moz-letter-spacing: 0.0025em;
  -ms-letter-spacing: 0.0025em;
  letter-spacing: 0.0025em;
  display: block;
}

.c1 {
  padding: 16px;
  background-color: white;
  border-radius: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 .+ .{
  margin-left: 16px;
}

.c1 .c4 + .c6 {
  margin-top: 3px;
}

.c2 {
  width: 40px;
  height: 40px;
  border-radius: 50%;
  background-image: linear-gradient(180deg,#0D658A 0%,#B0FFB4 100%);
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  width: 21px;
  height: 21px;
  -webkit-transform: translate(-1px,0px);
  -ms-transform: translate(-1px,0px);
  transform: translate(-1px,0px);
}

.c0 {
  -webkit-animation: 500ms linear ehxozz;
  animation: 500ms linear ehxozz;
}

<div
  className="c0"
  type="tremble"
>
  <div
    className="c1"
  >
    <div
      className=""
    >
      <div
        className="c2"
        type="deadCell"
      >
        <img
          alt="skull"
          className="c3"
          height="auto"
          src="skull.png"
          type="deadCell"
          width="21px"
        />
      </div>
    </div>
    <div
      className=""
    >
      <h6
        className="c4 c5"
      >
        Мёртвая
      </h6>
      <span
        className="c6 c7"
      >
        или прикидывается
      </span>
    </div>
  </div>
</div>
`;

exports[`Basic visual component should be rendered correctly Life component valid snapshot 1`] = `
.c5 {
  font-weight: 500;
  font-size: 20px;
  margin: 0;
  -webkit-letter-spacing: 0.0015em;
  -moz-letter-spacing: 0.0015em;
  -ms-letter-spacing: 0.0015em;
  letter-spacing: 0.0015em;
}

.c7 {
  -webkit-letter-spacing: 0.0025em;
  -moz-letter-spacing: 0.0025em;
  -ms-letter-spacing: 0.0025em;
  letter-spacing: 0.0025em;
  display: block;
}

.c1 {
  padding: 16px;
  background-color: white;
  border-radius: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 .+ .{
  margin-left: 16px;
}

.c1 .c4 + .c6 {
  margin-top: 3px;
}

.c2 {
  width: 40px;
  height: 40px;
  border-radius: 50%;
  background-image: linear-gradient(180deg,#AD00FF 0%,#FFB0E9 100%);
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c3 {
  width: 21px;
  height: 21px;
  -webkit-transform: translate(-1px,0px);
  -ms-transform: translate(-1px,0px);
  transform: translate(-1px,0px);
}

.c0 {
  -webkit-animation: 400ms ease-in ffhdPe;
  animation: 400ms ease-in ffhdPe;
}

<div
  className="c0"
  type="fadeId"
>
  <div
    className="c1"
  >
    <div
      className=""
    >
      <div
        className="c2"
        type="life"
      >
        <img
          alt="chicken"
          className="c3"
          height="auto"
          src="chicken.png"
          type="life"
          width="21px"
        />
      </div>
    </div>
    <div
      className=""
    >
      <h6
        className="c4 c5"
      >
        Жизнь
      </h6>
      <span
        className="c6 c7"
      >
        Ку-ку!
      </span>
    </div>
  </div>
</div>
`;
